chain/chunks/src/lib.rs

```diff
@@ -1091,7 +1124,80 @@ impl ShardsManager {
             return Ok(ProcessPartialEncodedChunkResult::HaveAllPartsAndReceipts(prev_block_hash));
         }
 
-        Ok(ProcessPartialEncodedChunkResult::NeedMorePartsOrReceipts(Box::new(header)))
+        match self.chunk_forwards_cache.cache_remove(&chunk_hash) {
+            None => Ok(ProcessPartialEncodedChunkResult::NeedMorePartsOrReceipts(Box::new(header))),
+
+            Some(forwarded_parts) => {
+                // We have the header now, and there were some parts we were forwarded earlier.
+                // Let's process those parts now.
+                let forwarded_chunk = PartialEncodedChunk {
+                    header: header.clone(),
+                    parts: forwarded_parts.into_iter().map(|(_, part)| part).collect(),
+                    receipts: Vec::new(),
+                };
+                // Make sure we mark this chunk as being requested so that it is handled
+                // properly in the next call. Note no requests are actually sent at this point.
```

Comment:
Not sure what "handled properly" means here. Can you elaborate on that?

Reply:
It's about this block [here](https://github.com/nearprotocol/nearcore/blob/master/chain/chunks/src/lib.rs#L950) which checks for duplicate chunks at the same height. The block is guarded by first checking whether we requested this chunk or not. By marking the chunk as requested here it means we will not view the forwarded parts as duplicate.

Reply:
I see.